### PR TITLE
Fixes a bug that prevent PartitionHiveParquetScroogeSource from being able to read files with the correct path.

### LINKED ISSUE #82: PartitionHiveParquetScroogeSource can't read from data written by PartitionHiveParquetScroogeSink
See
https://github.com/CommBank/bureaucracy/commit/fc6c4a8e4505dd01ad68c8e8d9e0bbf313487082#commitcomment-9790295
for some more background.

### PR DESCRIPTION
**Pull request by stephanh**
This closes #82 and supersedes #83.

### CODE REVIEW

File: hive/src/main/scala/au/com/cba/omnia/ebenezer/scrooge/hive/ParititionHiveParquetTap.scala [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+
+package au.com.cba.omnia.ebenezer.scrooge.hive
+
+import org.apache.hadoop.mapred.JobConf
+
+import cascading.scheme.Scheme
+import cascading.tap.{Tap, SinkMode}
+
```
Comment (vineethvarghese) [inline hive/src/main/scala/au/com/cba/omnia/ebenezer/scrooge/hive/ParititionHiveParquetTap.scala]: Space

Comment (stephanh) [inline hive/src/main/scala/au/com/cba/omnia/ebenezer/scrooge/hive/ParititionHiveParquetTap.scala]: Sorry I can't see it. What is wrong?

Comment (vineethvarghese) [inline hive/src/main/scala/au/com/cba/omnia/ebenezer/scrooge/hive/ParititionHiveParquetTap.scala]: line 21. Between

import cascading.tap.{Tap, SinkMode}

import cascading.tap.hive.{HiveTap, HivePartitionTap, HiveTableDescriptor}

Does cascading hive come from a different library? If it does, then ignore that comment.

Comment (stephanh) [inline hive/src/main/scala/au/com/cba/omnia/ebenezer/scrooge/hive/ParititionHiveParquetTap.scala]: Thank you. I was just blind :smile:.

File: hive/src/main/scala/au/com/cba/omnia/ebenezer/scrooge/hive/PartitionHiveParquetScroogeSource.scala [hunk 2]
```diff
@@ -188,7 +158,7 @@ case class PartitionHiveParquetScroogeSource[T <: ThriftStruct](
     HadoopSchemeInstance(Util.parquetScheme[T])
 
   override def createTap(readOrWrite: AccessMode)(implicit mode: Mode): Tap[_, _, _] = mode match {
-    case Local(_)              => sys.error("Local mode is currently not supported for ${toString}")
+    case Local(_) => sys.error("Local mode is currently not supported for ${toString}")
```
Comment (SamRoberts) [inline hive/src/main/scala/au/com/cba/omnia/ebenezer/scrooge/hive/PartitionHiveParquetScroogeSource.scala]: Minor comment: isn't this meant to be vertically aligned with the `=>` below?

Comment (stephanh) [inline hive/src/main/scala/au/com/cba/omnia/ebenezer/scrooge/hive/PartitionHiveParquetScroogeSource.scala]: I rearranged things to make them more readable but some of aligning doesn't actually improve readability across the multiline case blocks.

File: hive/src/main/scala/au/com/cba/omnia/ebenezer/scrooge/hive/ParititionHiveParquetTap.scala [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+}
+
+/** Provides utility functions for the PartitionHiveParquetTaps. */
+object PartitionHiveParquetTap {
+  /** Creates a path glob for all the files in a partitioned hive table. */
+  def createGlobbedPath(path: String, tableDescriptor: HiveTableDescriptor) =
+    s"$path/${partitionColumnsGlob(tableDescriptor)}"
+
+  /** Creates globs (`*`) for all the partition directories. */
+  def partitionColumnsGlob(tableDescriptor: HiveTableDescriptor) =
```
Comment (SamRoberts) [inline hive/src/main/scala/au/com/cba/omnia/ebenezer/scrooge/hive/ParititionHiveParquetTap.scala]: minor: Is it really worth separating out the two little pieces of functionality in createGlobbedPath and partitionColumnsGlob

File: hive/src/main/scala/au/com/cba/omnia/ebenezer/scrooge/hive/PartitionHiveParquetScroogeSource.scala [hunk 4]
```diff
@@ -77,9 +77,7 @@ case class PartitionHiveParquetScroogeSink[A, T <: ThriftStruct](
         val tableDescriptor =
           Util.createHiveTableDescriptor[T](database, table, partitionColumns, ParquetFormat, path)
 
-        val tap = new HivePartitionTap(
-          new HiveTap(tableDescriptor, hdfsScheme, SinkMode.REPLACE, true), SinkMode.UPDATE
-        )
+        val tap = new PartitionHiveParquetWriteTap(tableDescriptor, hdfsScheme)
```
Comment (vineethvarghese) [inline hive/src/main/scala/au/com/cba/omnia/ebenezer/scrooge/hive/PartitionHiveParquetScroogeSource.scala]: Minor point. I was confused a bit to see a WriteTap for Source. Upon closer inspection I see that both source and sink are in a file called *Source.scala :)

**Comment by SamRoberts** [comment]
:+1:

**Comment by vineethvarghese** [comment]
:+1:

File: hive/src/main/scala/au/com/cba/omnia/ebenezer/scrooge/hive/ParititionHiveParquetTap.scala (hunk 3 above)
Comment (RayRoestenburg) [inline hive/src/main/scala/au/com/cba/omnia/ebenezer/scrooge/hive/ParititionHiveParquetTap.scala]: I separated it since it's used in two places (dry).

Comment (SamRoberts) [inline hive/src/main/scala/au/com/cba/omnia/ebenezer/scrooge/hive/ParititionHiveParquetTap.scala]: `partitionColumnsGlob` is only used by `createGlobbedPath` that I can see?

Comment (RayRoestenburg) [inline hive/src/main/scala/au/com/cba/omnia/ebenezer/scrooge/hive/ParititionHiveParquetTap.scala]: Ah I misunderstood what you meant. I thought you were talking about the reason for the PartitionHiveParquetTap object (separating out the functions it provides to the two taps).

Comment (SamRoberts) [inline hive/src/main/scala/au/com/cba/omnia/ebenezer/scrooge/hive/ParititionHiveParquetTap.scala]: Ah, yes, I can see how you thought that. I should have been clearer!

Comment (RayRoestenburg) [inline hive/src/main/scala/au/com/cba/omnia/ebenezer/scrooge/hive/ParititionHiveParquetTap.scala]: no problem :-)

**Comment by RayRoestenburg** [comment]
There is an eviction warning in the build, not sure how important it is to cleanup this dependency?

        [warn] 	* com.twitter:scrooge-core_2.10:3.6.0 -> 3.14.1

**Comment by SamRoberts** [comment]
Yeah, I am looking at dependencies. (Or at least, I was, and I will start again tomorrow.) This is one of many we have to look at.

**Comment by RayRoestenburg** [comment]
@SamRoberts ah great. I've got some dependency issues in my current project which I will hope will disappear once you're done ;-)

**Comment by RayRoestenburg** [comment]
@SamRoberts FYI as for dependencies, noticed that bijection-scrooge and scrooge-core in uniform cause the eviction warning, bijection-scrooge depends on scrooge-core 3.6.0 (through serializer, runtime) while uniform defines 3.14.1 for scrooge-core.

**Pull request merged**